src/renderers/dom/shared/__tests__/ReactDOMAttribute-test.js

```diff
@@ -0,0 +1,110 @@
+ *
+ * @emails react-core
+ */
+
+'use strict';
+
+describe('ReactDOM unknown attribute', () => {
+  var React;
+  var ReactDOM;
+
+  beforeEach(() => {
+    jest.resetModules();
+    React = require('react');
+    ReactDOM = require('react-dom');
+  });
+
+  function testUnknownAttributeRemoval(givenValue) {
+    var el = document.createElement('div');
+    ReactDOM.render(<div unknown="something" />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe('something');
+    ReactDOM.render(<div unknown={givenValue} />, el);
+    expect(el.firstChild.hasAttribute('unknown')).toBe(false);
+  }
+
+  function testUnknownAttributeAssignment(givenValue, expectedDOMValue) {
+    var el = document.createElement('div');
+    ReactDOM.render(<div unknown="something" />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe('something');
+    ReactDOM.render(<div unknown={givenValue} />, el);
+    expect(el.firstChild.getAttribute('unknown')).toBe(expectedDOMValue);
+  }
+
+  describe('unknown attributes', () => {
+    it('removes values null and undefined', () => {
+      testUnknownAttributeRemoval(null);
+      testUnknownAttributeRemoval(undefined);
+    });
+
+    it('removes unknown attributes that were rendered but are now missing', () => {
+      var el = document.createElement('div');
+      ReactDOM.render(<div unknown="something" />, el);
+      expect(el.firstChild.getAttribute('unknown')).toBe('something');
+      ReactDOM.render(<div />, el);
+      expect(el.firstChild.hasAttribute('unknown')).toBe(false);
+    });
+
+    it('passes through strings', () => {
+      testUnknownAttributeAssignment('a string', 'a string');
+    });
+
+    it('coerces numbers and booleans to strings', () => {
+      testUnknownAttributeAssignment(0, '0');
+      testUnknownAttributeAssignment(-1, '-1');
+      testUnknownAttributeAssignment(42, '42');
+      testUnknownAttributeAssignment(9000.99, '9000.99');
+      testUnknownAttributeAssignment(true, 'true');
+      testUnknownAttributeAssignment(false, 'false');
+    });
+
+    it('coerces NaN to strings **and warns**', () => {
+      spyOn(console, 'error');
+
+      testUnknownAttributeAssignment(NaN, 'NaN');
+      // TODO: add specific expectations about what the warning says
+      // expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(...
+      expectDev(console.error.calls.count()).toBe(1);
+    });
+
+    it('coerces objects to strings **and warns**', () => {
+      spyOn(console, 'error');
+
+      const lol = {
+        toString() {
+          return 'lol';
+        },
+      };
+
+      testUnknownAttributeAssignment({hello: 'world'}, '[object Object]');
+      testUnknownAttributeAssignment(lol, 'lol');
+      // TODO: add specific expectations about what the warning says
+      // expectDev(normalizeCodeLocInfo(console.error.calls.argsFor(0)[0])).toBe(...
+      expectDev(console.error.calls.count()).toBe(1);
```

Comment:
Hmm that's what I recorded in my notes after yesterday's conversation, but I may be mistaken. I though the rationale was that `toString` could be slow and we should encourage people not to rely on it.

@sebmarkbage